storage/core/src/util.rs

```diff
@@ -0,0 +1,42 @@
+// Copyright (c) 2022 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Utilities for implementing storage backends
+
+use crate::Data;
+use std::collections::BTreeMap;
+
+/// Iterator over entries of a [BTreeMap] with keas starting with given prefix
+pub struct PrefixIter<'m, T> {
+    inner: std::collections::btree_map::Range<'m, Data, T>,
+    prefix: Data,
+}
+
+impl<'m, T> PrefixIter<'m, T> {
+    pub fn new(map: &'m BTreeMap<Data, T>, prefix: Data) -> Self {
```

Comment:
I did experiment with this. Here are some high-level observations:

Taking prefix by reference `&[u8]` is not quite trivial but achievable:

* The lifetime of the prefix has to be tied to the lifetime of the iterator. That is somewhat less convenient for the user because the burden of making sure the prefix data live for at least as long as the iterator is placed on the user.
* Due to how the `BTreeMap` range iterator works, an extra wrapper type with custom implementation of `std::ops::RangeBounds` is required to match the number of indirections.

Also note this is the low-level storage interface. It is accessed by the high-level interface, which operates by encoding some object and passing it into the `prefix_iter`, and returning a new iterator. For this use case, the owned value is generated anyway, and it has to be passed out of the function. The most straightforward way to do it is to include the prefix in the iterator as an owned value. Reference alone does not work, since the prefix is created inside the function but it has to outlive the function scope.

Clearly, some use cases are better served by `&[u8]` and  some by `Vec<u8>`. In general, it would make sense to parametrize the iterator over type that can give prefix slice, i.e. instead of hard-coding `Vec<u8>` or `&[u8]` make it polymorphic over `T: Borrow<[u8]>`. The `BTreeMap` range iterator does something like this. However, given that all calls ultimately go through the high-level interface which uses the owned variant, I don't see much benefit doing it at present.

Reply:
Alrighty. Let's go for it without this for now, but maybe open an issue to see the final outcome later and whether we want to pursue this based on it, and link to this comment.

Reply:
> Alrighty. Let's go for it without this for now, but maybe open an issue to see the final outcome later and whether we want to pursue this based on it, and link to this comment.

Issue #417